Verification workflow is failing on centos7 and s390x

On centos7 the error is:

```
  /nanoarrow/src/nanoarrow/common/array_test.cc: In member function ‘virtual void ArrayViewTest_ArrayViewTestGetStringView_Test::TestBody()’:
  /nanoarrow/src/nanoarrow/common/array_test.cc:3486:48: error: ‘StringViewBuilder’ was not declared in this scope
     auto string_view_builder = StringViewBuilder();
                                                  ^
```

https://github.com/apache/arrow-nanoarrow/actions/runs/11088204305/job/30807816230#step:4:405

...and on s390x the error is:

```
Note: Google Test filter = NanoarrowIpcTest.NanoarrowIpcCheckHeader
  [==========] Running 1 test from 1 test suite.
  [----------] Global test environment set-up.
  [----------] 1 test from NanoarrowIpcTest
  [ RUN      ] NanoarrowIpcTest.NanoarrowIpcCheckHeader
  /nanoarrow/src/nanoarrow/ipc/decoder_test.cc:135: Failure
  Expected equality of these values:
    ArrowIpcDecoderVerifyHeader(&decoder, data, &error)
      Which is: 29
    22
  
  /nanoarrow/src/nanoarrow/ipc/decoder_test.cc:136: Failure
  Expected equality of these values:
    error.message
      Which is: "Expected >= 16777219 bytes of remaining data but found 8 bytes in buffer"
    "Expected 0xFFFFFFFF at start of message but found 0xFFFFFF00"
  
  /nanoarrow/src/nanoarrow/ipc/decoder_test.cc:140: Failure
  Expected equality of these values:
    ArrowIpcDecoderDecodeHeader(&decoder, data, &error)
      Which is: 29
    22
  
  /nanoarrow/src/nanoarrow/ipc/decoder_test.cc:141: Failure
  Expected equality of these values:
    error.message
      Which is: "Expected >= 16777219 bytes of remaining data but found 8 bytes in buffer"
    "Expected 0xFFFFFFFF at start of message but found 0xFFFFFF00"
```

https://github.com/apache/arrow-nanoarrow/actions/runs/11088204305/job/30807816936#step:4:817

...which I think just means we need to expect a different error message on big endian.

--- a/r/tests/testthat/test-ipc.R
+++ b/r/tests/testthat/test-ipc.R
@@ -103,7 +103,7 @@ test_that("write_nanoarrow() works for file paths", {
   tf <- tempfile()
   on.exit(unlink(tf))
 
-  df <- data.frame(letters = letters)
+  df <- data.frame(letters = letters, stringsAsFactors = FALSE)
   expect_identical(write_nanoarrow(df, tf), df)
   expect_identical(as.data.frame(read_nanoarrow(tf)), df)
 })
@@ -127,7 +127,7 @@ test_that("write_nanoarrow() works for URLs", {
   tf <- tempfile()
   on.exit(unlink(tf))
 
-  df <- data.frame(letters = letters)
+  df <- data.frame(letters = letters, stringsAsFactors = FALSE)
   expect_identical(write_nanoarrow(df, paste0("file://", tf)), df)
   expect_identical(as.data.frame(read_nanoarrow(tf)), df)
 })
@@ -151,7 +151,7 @@ test_that("write_nanoarrow() works for compressed .gz file paths", {
   tf <- tempfile(fileext = ".gz")
   on.exit(unlink(tf))
 
-  df <- data.frame(letters = letters)
+  df <- data.frame(letters = letters, stringsAsFactors = FALSE)
   expect_identical(write_nanoarrow(df, tf), df)
   expect_identical(as.data.frame(read_nanoarrow(tf)), df)
 })
@@ -175,7 +175,7 @@ test_that("write_nanoarrow() works for compressed .bz2 file paths", {
   tf <- tempfile(fileext = ".bz2")
   on.exit(unlink(tf))
 
-  df <- data.frame(letters = letters)
+  df <- data.frame(letters = letters, stringsAsFactors = FALSE)
   expect_identical(write_nanoarrow(df, tf), df)
   expect_identical(as.data.frame(read_nanoarrow(tf)), df)
 })
@@ -212,7 +212,7 @@ test_that("write_nanoarrow() errors for compressed .zip file paths", {
   tf <- tempfile(fileext = ".zip")
   on.exit(unlink(tf))
 
-  df <- data.frame(letters = letters)
+  df <- data.frame(letters = letters, stringsAsFactors = FALSE)
   expect_error(
     write_nanoarrow(df, tf),
     "zip compression not supported"

--- a/src/nanoarrow/common/array_test.cc
+++ b/src/nanoarrow/common/array_test.cc
@@ -895,12 +895,12 @@ TEST(ArrayTest, ArrayTestAppendToLargeStringArray) {
   ArrowArrayRelease(&array);
 }
 
-template <enum ArrowType ArrowT, typename ValueT>
 void TestAppendToInlinedDataViewArray(
-    std::function<ArrowErrorCode(struct ArrowArray*, ValueT)> AppendFunc) {
+    enum ArrowType arrow_type,
+    std::function<ArrowErrorCode(struct ArrowArray*, const char*)> AppendFunc) {
   struct ArrowArray array;
 
-  ASSERT_EQ(ArrowArrayInitFromType(&array, ArrowT), NANOARROW_OK);
+  ASSERT_EQ(ArrowArrayInitFromType(&array, arrow_type), NANOARROW_OK);
   EXPECT_EQ(ArrowArrayStartAppending(&array), NANOARROW_OK);
 
   // Check that we can reserve
@@ -908,7 +908,7 @@ void TestAppendToInlinedDataViewArray(
   EXPECT_EQ(ArrowArrayBuffer(&array, 1)->capacity_bytes,
             5 * sizeof(union ArrowBinaryView));
 
-  EXPECT_EQ(AppendFunc(&array, ValueT{{"inlinestring"}, 12}), NANOARROW_OK);
+  EXPECT_EQ(AppendFunc(&array, "inlinestring"), NANOARROW_OK);
   EXPECT_EQ(ArrowArrayAppendNull(&array, 2), NANOARROW_OK);
   EXPECT_EQ(ArrowArrayAppendEmpty(&array, 1), NANOARROW_OK);
   EXPECT_EQ(ArrowArrayFinishBuildingDefault(&array, nullptr), NANOARROW_OK);
@@ -934,12 +934,12 @@ void TestAppendToInlinedDataViewArray(
   ArrowArrayRelease(&array);
 };
 
-template <enum ArrowType ArrowT, typename ValueT>
 void TestAppendToDataViewArray(
-    std::function<ArrowErrorCode(struct ArrowArray*, ValueT)> AppendFunc) {
+    enum ArrowType arrow_type,
+    std::function<ArrowErrorCode(struct ArrowArray*, const char*)> AppendFunc) {
   struct ArrowArray array;
 
-  ASSERT_EQ(ArrowArrayInitFromType(&array, ArrowT), NANOARROW_OK);
+  ASSERT_EQ(ArrowArrayInitFromType(&array, arrow_type), NANOARROW_OK);
   EXPECT_EQ(ArrowArrayStartAppending(&array), NANOARROW_OK);
 
   // Check that we can reserve
@@ -951,15 +951,12 @@ void TestAppendToDataViewArray(
   std::string filler(NANOARROW_BINARY_VIEW_BLOCK_SIZE - 34, 'x');
   std::string str2{"goes_into_second_variadic_buffer"};
 
-  EXPECT_EQ(AppendFunc(&array, ValueT{{"1234"}, 4}), NANOARROW_OK);
+  EXPECT_EQ(AppendFunc(&array, "1234"), NANOARROW_OK);
   EXPECT_EQ(ArrowArrayAppendNull(&array, 2), NANOARROW_OK);
-  EXPECT_EQ(AppendFunc(&array, {{str1.c_str()}, static_cast<int64_t>(str1.size())}),
-            NANOARROW_OK);
+  EXPECT_EQ(AppendFunc(&array, str1.c_str()), NANOARROW_OK);
   EXPECT_EQ(ArrowArrayAppendEmpty(&array, 1), NANOARROW_OK);
-  EXPECT_EQ(AppendFunc(&array, {{filler.c_str()}, static_cast<int64_t>(filler.size())}),
-            NANOARROW_OK);
-  EXPECT_EQ(AppendFunc(&array, {{str2.c_str()}, static_cast<int64_t>(str2.size())}),
-            NANOARROW_OK);
+  EXPECT_EQ(AppendFunc(&array, filler.c_str()), NANOARROW_OK);
+  EXPECT_EQ(AppendFunc(&array, str2.c_str()), NANOARROW_OK);
   EXPECT_EQ(ArrowArrayFinishBuildingDefault(&array, nullptr), NANOARROW_OK);
 
   EXPECT_EQ(array.length, 7);
@@ -1004,17 +1001,28 @@ void TestAppendToDataViewArray(
 };
 
 TEST(ArrayTest, ArrayTestAppendToBinaryViewArray) {
-  TestAppendToInlinedDataViewArray<NANOARROW_TYPE_STRING_VIEW, struct ArrowStringView>(
-      ArrowArrayAppendString);
-  TestAppendToDataViewArray<NANOARROW_TYPE_STRING_VIEW, struct ArrowStringView>(
-      ArrowArrayAppendString);
+  TestAppendToInlinedDataViewArray(
+      NANOARROW_TYPE_STRING_VIEW, [&](struct ArrowArray* array, const char* value) {
+        return ArrowArrayAppendString(array, ArrowCharView(value));
+      });
+
+  TestAppendToDataViewArray(NANOARROW_TYPE_STRING_VIEW,
+                            [&](struct ArrowArray* array, const char* value) {
+                              return ArrowArrayAppendString(array, ArrowCharView(value));
+                            });
 };
 
 TEST(ArrayTest, ArrayTestAppendToStringViewArray) {
-  TestAppendToInlinedDataViewArray<NANOARROW_TYPE_BINARY_VIEW, struct ArrowBufferView>(
-      ArrowArrayAppendBytes);
-  TestAppendToDataViewArray<NANOARROW_TYPE_BINARY_VIEW, struct ArrowBufferView>(
-      ArrowArrayAppendBytes);
+  TestAppendToInlinedDataViewArray(
+      NANOARROW_TYPE_BINARY_VIEW, [&](struct ArrowArray* array, const char* value) {
+        return ArrowArrayAppendBytes(array,
+                                     {{value}, static_cast<int64_t>(strlen(value))});
+      });
+
+  TestAppendToDataViewArray(NANOARROW_TYPE_BINARY_VIEW, [&](struct ArrowArray* array,
+                                                            const char* value) {
+    return ArrowArrayAppendBytes(array, {{value}, static_cast<int64_t>(strlen(value))});
+  });
 };
 
 TEST(ArrayTest, ArrayTestAppendToFixedSizeBinaryArray) {
@@ -3562,15 +3570,20 @@ void TestGetFromBinaryView(
 }
 
 TEST(ArrayViewTest, ArrayViewTestGetStringView) {
+#if defined(ARROW_VERSION_MAJOR) && ARROW_VERSION_MAJOR >= 15
   auto string_view_builder = StringViewBuilder();
   const auto get_string_view = [](const struct ArrowStringView* sv) { return sv->data; };
   TestGetFromInlinedBinaryView<StringViewBuilder, struct ArrowStringView>(
       string_view_builder, ArrowArrayViewGetStringUnsafe, get_string_view);
   TestGetFromBinaryView<StringViewBuilder, struct ArrowStringView>(
       string_view_builder, ArrowArrayViewGetStringUnsafe, get_string_view);
+#else
+  GTEST_SKIP() << "Arrow C++ StringView compatibility test needs Arrow C++ >= 15";
+#endif
 }
 
 TEST(ArrayViewTest, ArrayViewTestGetBinaryView) {
+#if defined(ARROW_VERSION_MAJOR) && ARROW_VERSION_MAJOR >= 15
   auto binary_view_builder = BinaryViewBuilder();
   const auto get_buffer_view = [](const struct ArrowBufferView* bv) {
     return bv->data.data;
@@ -3579,6 +3592,9 @@ TEST(ArrayViewTest, ArrayViewTestGetBinaryView) {
       binary_view_builder, ArrowArrayViewGetBytesUnsafe, get_buffer_view);
   TestGetFromBinaryView<BinaryViewBuilder, struct ArrowBufferView>(
       binary_view_builder, ArrowArrayViewGetBytesUnsafe, get_buffer_view);
+#else
+  GTEST_SKIP() << "Arrow C++ StringView compatibility test needs Arrow C++ >= 15";
+#endif
 }
 
 TEST(ArrayViewTest, ArrayViewTestGetIntervalYearMonth) {

--- a/src/nanoarrow/common/array_test.cc
+++ b/src/nanoarrow/common/array_test.cc
@@ -1002,24 +1002,23 @@ void TestAppendToDataViewArray(
 
 TEST(ArrayTest, ArrayTestAppendToBinaryViewArray) {
   TestAppendToInlinedDataViewArray(
-      NANOARROW_TYPE_STRING_VIEW, [&](struct ArrowArray* array, const char* value) {
+      NANOARROW_TYPE_STRING_VIEW, [&](ArrowArray* array, const char* value) {
         return ArrowArrayAppendString(array, ArrowCharView(value));
       });
 
   TestAppendToDataViewArray(NANOARROW_TYPE_STRING_VIEW,
-                            [&](struct ArrowArray* array, const char* value) {
+                            [&](ArrowArray* array, const char* value) {
                               return ArrowArrayAppendString(array, ArrowCharView(value));
                             });
 };
 
 TEST(ArrayTest, ArrayTestAppendToStringViewArray) {
-  TestAppendToInlinedDataViewArray(
-      NANOARROW_TYPE_BINARY_VIEW, [&](struct ArrowArray* array, const char* value) {
-        return ArrowArrayAppendBytes(array,
-                                     {{value}, static_cast<int64_t>(strlen(value))});
-      });
+  TestAppendToInlinedDataViewArray(NANOARROW_TYPE_BINARY_VIEW, [&](ArrowArray* array,
+                                                                   const char* value) {
+    return ArrowArrayAppendBytes(array, {{value}, static_cast<int64_t>(strlen(value))});
+  });
 
-  TestAppendToDataViewArray(NANOARROW_TYPE_BINARY_VIEW, [&](struct ArrowArray* array,
+  TestAppendToDataViewArray(NANOARROW_TYPE_BINARY_VIEW, [&](ArrowArray* array,
                                                             const char* value) {
     return ArrowArrayAppendBytes(array, {{value}, static_cast<int64_t>(strlen(value))});
   });

--- a/src/nanoarrow/common/schema_test.cc
+++ b/src/nanoarrow/common/schema_test.cc
@@ -912,7 +912,13 @@ TEST(SchemaViewTest, SchemaViewInitBinaryAndString) {
   EXPECT_EQ(schema_view.layout.element_size_bits[2], 0);
   EXPECT_EQ(ArrowSchemaToStdString(&schema), "large_string");
   ArrowSchemaRelease(&schema);
+}
 
+TEST(SchemaViewTest, SchemaViewInitBinaryAndStringView) {
+#if defined(ARROW_VERSION_MAJOR) && ARROW_VERSION_MAJOR >= 15
+  struct ArrowSchema schema;
+  struct ArrowSchemaView schema_view;
+  struct ArrowError error;
   ARROW_EXPECT_OK(ExportType(*utf8_view(), &schema));
   EXPECT_EQ(ArrowSchemaViewInit(&schema_view, &schema, &error), NANOARROW_OK);
   EXPECT_EQ(schema_view.type, NANOARROW_TYPE_STRING_VIEW);
@@ -938,6 +944,9 @@ TEST(SchemaViewTest, SchemaViewInitBinaryAndString) {
   EXPECT_EQ(schema_view.layout.element_size_bits[1], 128);
   EXPECT_EQ(ArrowSchemaToStdString(&schema), "binary_view");
   ArrowSchemaRelease(&schema);
+#else
+  GTEST_SKIP() << "Arrow C++ StringView compatibility test needs Arrow C++ >= 15";
+#endif
 }
 
 TEST(SchemaViewTest, SchemaViewInitBinaryAndStringErrors) {

--- a/src/nanoarrow/ipc/decoder_test.cc
+++ b/src/nanoarrow/ipc/decoder_test.cc
@@ -132,14 +132,30 @@ TEST(NanoarrowIpcTest, NanoarrowIpcCheckHeader) {
   uint32_t eight_bad_bytes[] = {negative_one_le * 256, 999};
   data.data.as_uint32 = eight_bad_bytes;
   data.size_bytes = sizeof(eight_bad_bytes);
+
+#if defined(__BIG_ENDIAN__)
+  EXPECT_EQ(ArrowIpcDecoderVerifyHeader(&decoder, data, &error), ESPIPE);
+  EXPECT_STREQ(
+      error.message,
+      "Expected >= 16777219 bytes of remaining data but found 8 bytes in buffer");
+#else
   EXPECT_EQ(ArrowIpcDecoderVerifyHeader(&decoder, data, &error), EINVAL);
   EXPECT_STREQ(error.message,
                "Expected 0xFFFFFFFF at start of message but found 0xFFFFFF00");
+#endif
 
   ArrowErrorInit(&error);
+
+#if defined(__BIG_ENDIAN__)
+  EXPECT_EQ(ArrowIpcDecoderDecodeHeader(&decoder, data, &error), ESPIPE);
+  EXPECT_STREQ(
+      error.message,
+      "Expected >= 16777219 bytes of remaining data but found 8 bytes in buffer");
+#else
   EXPECT_EQ(ArrowIpcDecoderDecodeHeader(&decoder, data, &error), EINVAL);
   EXPECT_STREQ(error.message,
                "Expected 0xFFFFFFFF at start of message but found 0xFFFFFF00");
+#endif
 
   eight_bad_bytes[0] = 0xFFFFFFFF;
   eight_bad_bytes[1] = negative_one_le;

